scripts/paragraph_segmentation_dcnn.py

```diff
@@ -0,0 +1,161 @@
+
+MODEL_CHECKPOINT_FOLDER = "model_checkpoint"
+if not os.path.isdir(MODEL_CHECKPOINT_FOLDER):
+    os.makedirs(MODEL_CHECKPOINT_FOLDER)
+
+train_ds = IAMDataset("form", output_data="bb", output_parse_method="form", train=True)
+print("Number of training samples: {}".format(len(train_ds)))
+
+test_ds = IAMDataset("form", output_data="bb", output_parse_method="form", train=False)
+print("Number of testing samples: {}".format(len(test_ds)))
+
+batch_size = 32
+
+def transform(data, label):
+    image = mx.nd.array(data).expand_dims(axis=2)
+    image = resize_short(image, int(800/3))
+    image = image.transpose([2, 0, 1])/255.
+    label = label[0].astype(np.float32)
+    return image, mx.nd.array(label)
+
+def augment_transform(data, label):
+    width_range = 0.05
+    height_range = 0.05
+
+    ty = random.uniform(-height_range, height_range)
+    tx = random.uniform(-width_range, width_range)
+    st = tf.SimilarityTransform(translation=(tx*data.shape[1], ty*data.shape[0]))
+    data = tf.warp(data, st)
+
+    label[0][0] = label[0][0] - tx
+    label[0][1] = label[0][1] - ty
+    return transform(data*255., label)
+    
+train_data = gluon.data.DataLoader(train_ds.transform(augment_transform), batch_size, shuffle=True, num_workers=multiprocessing.cpu_count())
+test_data = gluon.data.DataLoader(test_ds.transform(transform), batch_size, shuffle=False, num_workers=multiprocessing.cpu_count())
+
+def make_cnn():
+    p_dropout = 0.5
+
+    cnn = gluon.nn.HybridSequential()
+    cnn.add(gluon.nn.Conv2D(kernel_size=(3,3), padding=(1,1), channels=16, activation="relu"))
+    cnn.add(gluon.nn.BatchNorm())
+
+    cnn.add(gluon.nn.Conv2D(kernel_size=(3,3), padding=(1,1), channels=16, activation="relu"))
+    cnn.add(gluon.nn.BatchNorm())
+
+    cnn.add(gluon.nn.Conv2D(kernel_size=(3,3), padding=(1,1), channels=16, activation="relu"))
+    cnn.add(gluon.nn.BatchNorm())
+
+    cnn.add(gluon.nn.Conv2D(kernel_size=(3,3), padding=(1,1), channels=16, activation="relu"))
+    cnn.add(gluon.nn.MaxPool2D(pool_size=(2,2), strides=(2,2)))
+    cnn.add(gluon.nn.BatchNorm())
+
+    cnn.add(gluon.nn.Conv2D(kernel_size=(3,3), padding=(1,1), channels=16, activation="relu"))
+    cnn.add(gluon.nn.MaxPool2D(pool_size=(2,2), strides=(2,2)))
+    cnn.add(gluon.nn.BatchNorm())
+
+    cnn.add(gluon.nn.Flatten())
+    cnn.add(gluon.nn.Dense(64, activation='relu'))
+    cnn.add(gluon.nn.Dropout(p_dropout))
+    cnn.add(gluon.nn.Dense(64, activation='relu'))
+    cnn.add(gluon.nn.Dropout(p_dropout))
+    cnn.add(gluon.nn.Dense(4, activation='sigmoid'))
+
+    cnn.hybridize()
+    return cnn
+
+cnn = make_cnn()
+cnn.load_params("model_checkpoint/cnn300.params", ctx=ctx)
+
+iou_loss = IOU_loss()
+LEARNING_RATE = 0.0001
+trainer = gluon.Trainer(cnn.collect_params(), 'adam', {'learning_rate': LEARNING_RATE, })
+
+epochs = 500
+print_every_n = 20
+send_image_every_n = 20
+save_every_n = 100
+
+for e in range(epochs):    
+    loss = nd.zeros(1, ctx)
```

Comment:
make this a `run_epoch(net, dataloader, loss_function) -> loss` and simplify your loop